jobs/kpi-forecasting/kpi_forecasting/models/base_forecast.py

```diff
@@ -115,116 +59,206 @@ def _validate_forecast_df(self, forecast_df: pd.DataFrame) -> None:
-        summary_df["metric_collected_at"] = self.collected_at
+    def fit(self, observed_df: pd.DataFrame) -> None:
+        """Fit models across all segments for the data in observed_df
 
-        # add forecast model metadata columns
-        summary_df["forecast_start_date"] = self.start_date
-        summary_df["forecast_end_date"] = self.end_date
-        summary_df["forecast_trained_at"] = self.trained_at
-        summary_df["forecast_predicted_at"] = self.predicted_at
-        summary_df["forecast_parameters"] = self.metadata_params
+        Args:
+            observed_df (pd.DataFrame): data used to fit
+        """
+        print(f"Fitting {self.model_type} model.", flush=True)
+        # create list of models depending on whether there are segments or not
+        self._set_segment_models(observed_df)
+        for segment_model in self.segment_models:
+            print(segment_model["segment"])
+            model = segment_model["model"]
+            model._set_seed()
+            observed_subset = self.filter_data_to_segment(
+                observed_df, segment_model["segment"], segment_model["start_date"]
+            )
+            model.fit(observed_subset)
+        return self
+
+    def predict(self, dates_to_predict: pd.DataFrame) -> pd.DataFrame:
+        """Generates a prediction for each segment for the dates in dates_to_predict
 
-        self.summary_df = summary_df
+        Args:
+            dates_to_predict (pd.DataFrame): dataframe with a single column,
+            submission_date that is a string in `%Y-%m-%d` format
 
-        return self.summary_df
+        Returns:
+            pd.DataFrame: prediction across all segments
+        """
+        start_date = dates_to_predict["submission_date"].iloc[0]
+        end_date = dates_to_predict["submission_date"].iloc[-1]
+
+        print(f"Forecasting from {start_date} to {end_date}.", flush=True)
+        for segment_model in self.segment_models:
+            config_start_date = segment_model["start_date"]
+
+            if config_start_date and config_start_date > start_date:
+                dates_to_predict_segment = (
+                    dates_to_predict[
+                        dates_to_predict["submission_date"] >= config_start_date
+                    ]
+                    .reset_index(drop=True)
+                    .copy()
+                )
+            else:
+                dates_to_predict_segment = dates_to_predict.copy()
+
+            model = segment_model["model"]
+            model._set_seed()
+            predict_df = model.predict(dates_to_predict_segment)
+
+            # add segments on as columns
+            for column, value in segment_model["segment"].items():
+                predict_df[column] = value
+            predict_df["forecast_parameters"] = json.dumps(model._get_parameters())
+            segment_model["forecast"] = predict_df
+        self.forecast_list = [el["forecast"] for el in self.segment_models]
```

Comment:
Why not use `i` or `x` here? "el" seems nonstandard and confusing.
```suggestion
        self.forecast_list = [i["forecast"] for i in self.segment_models]
```

Reply:
For some reason I got into the habit of using "el" in list comprehensions,  changed to `x`